site/content/api/Device-Connection-API.md

```diff
@@ -0,0 +1,143 @@
+
+The body of the message SHOULD be empty and will be ignored if it is not.
+
+**Response Message Format**
+
+A response to a *set last known gateway for device* request contains the [Standard Response Properties]({{< relref "#standard-response-properties" >}}).
+
+The response message's *status* property may contain the following codes:
+
+| Code | Description |
+| :--- | :---------- |
+| *204* | OK, the last known gateway for the device was updated. |
+
+Implementors of this API may choose to check whether there is a device registered with the given *device_id* within the given *tenant_id* or a gateway registered with the given *gateway_id*. In such a case a *404* status code may be returned.  
+
+For status codes indicating an error (codes in the `400 - 499` range) the message body MAY contain a detailed description of the error that occurred.
+
+### Get Last Known Gateway for Device
+
+Clients use this command to *retrieve* the gateway that last acted on behalf of a given device.
+
+**Message Flow**
+
+
+The following sequence diagram illustrates the flow of messages involved in a *Client* retrieving the last known gateway for a Device.
+
+![Get Last Known Gateway for Device message flow](../get-last-gw_Success.png)
+
+
+**Request Message Format**
+
+The following table provides an overview of the properties a client needs to set on a message to retrieve the last known gateway for a device in addition to the [Standard Request Properties]({{< relref "#standard-request-properties" >}}).
+
+| Name        | Mandatory | Location                 | AMQP Type | Description |
+| :---------- | :-------: | :----------------------- | :-------- | :---------- |
+| *subject*   | yes       | *properties*             | *string*  | MUST be set to `get-last-gw`. |
+
+The body of the message SHOULD be empty and will be ignored if it is not.
+
+**Response Message Format**
+
+A response to a *get last known gateway for device* request contains the [Standard Response Properties]({{< relref "#standard-response-properties" >}}).
+
+The response message includes payload with a JSON object having the following properties:
+
+| Name             | Mandatory | JSON Type     | Description |
+| :--------------- | :-------: | :------------ | :---------- |
+| *gateway-id*     | *yes*     | *string*      | The ID of the last known gateway for the device. |
+| *update-date*    | *no*      | *string*      | The date that the information about the last known gateway for the device was last updated. The value MUST be an [ISO 8601 compliant *combined date and time representation in extended format*](https://en.wikipedia.org/wiki/ISO_8601#Combined_date_and_time_representations). |
+
+The response message's *status* property may contain the following codes:
+
+| Code | Description |
+| :--- | :---------- |
+| *200* | OK, the payload contains the gateway ID. |
+| *404* | Not Found, there is no last known gateway assigned to the device. |
```

Comment:
I assume this means: *we do not have any information for the given device ID on record*.
Is this also returned if there is no device with the given ID? Do we need to distinguish these cases?
Would it make sense to always return a response with *gateway-id* set to *device_id* if we do not have any information on record?

Reply:
> I assume this means: *we do not have any information for the given device ID on record*.

Yes.

> Is this also returned if there is no device with the given ID?

Yes, if such an optional check is performed.

> Do we need to distinguish these cases?

I don't think so.

> Would it make sense to always return a response with *gateway-id* set to *device_id* if we do not have any information on record?

I wouldn't do that, I would return a 404. 
For a device with a *single* gateway in its `via` device information entry, commands should by default be routed to that gateway, IMHO. This doesn't mean that the Device Connection API should return that gateway id as default - that would impose a dependency on the device registration API.
Instead, the `GatewayMapper` implementation should first invoke `assertRegistration`, get the list of `via` entries in its response (see #1281) and then invoke the `getLastKnownGateway` method of the Device Connection API. With no last known gateway for the device on record, a 404 response should be returned for that request and the `GatewayMapper` should choose the single `via` gateway as mapping target.

Reply:
sounds reasonable. The Gateway Mapper can then also use the information returned in the *assert Registration* response to verify if the GW returned by the Device Connection service is actually in the list of *allowed* gateways ...